test/decl/func/dynamic_self.swift

```diff
@@ -24,9 +24,9 @@ enum E0 {
 class C0 {
   func f() -> Self { } // okay
 
-  func g(_ ds: Self) { } // expected-error{{'Self' is only available in a protocol or as the result of a method in a class; did you mean 'C0'?}}{{16-20=C0}}
+  func g(_ ds: Self) { } // okay
```

Comment:
These are not covariant uses though. You need to perform a similar check to what's done for protocol requirements; check out ProtocolDecl::findProtocolSelfReferences().

Reply:
All of these examples are supported :)

Reply:
Yes but you need to *reject* these examples, among others:

```
func f(_: Self) -> () {}
func f(_: () -> Self) -> () {}
func f(_: (inout Self) -> ()) -> () {}

struct G<T> {}
func f() -> G<Self> {}
```

Take a look at the rule for protocols -- its important that we only allow covariant Self since otherwise type safety will be violated.

Reply:
> These are not covariant uses though. You need to perform a similar check to what's done for protocol requirements; check out ProtocolDecl::findProtocolSelfReferences().

in lib/AST/Decl.cpp, findProtocolSelfReferences returns SelfReferenceKind which is a struct. Then what does following do?
```
if (findProtocolSelfReferences(...)) {}
```
Is it always true or is it equivalent to something like following?
```
if (findProtocolSelfReferences(...) != SelfReferenceKind::None()) {}
```

Reply:
Yep.

https://github.com/apple/swift/blob/master/include/swift/AST/Decl.h#L3884-L3886

Reply:
(Slightly more idiomatic would be if it returned an OptionSet, but you don't have to change that.)